benchmark/httpapi.py

```diff
@@ -100,6 +112,104 @@ def delete(self, id):
+class TxMongoBackend(object):
+    """
+    The backend that uses txmongo driver to work with MongoDB.
+    """
+    def __init__(self, hostname="127.0.0.1", port=27017):
+        connection = MongoConnectionPool(host=hostname, port=port)
+        self.collection = connection.benchmark.results
+
+    def disconnect(self):
+        return self.collection.database.connection.disconnect()
+
+    def store(self, result):
+        """
+        Store a single benchmarking result and return its identifier.
+
+        :param dict result: The result in the JSON compatible format.
+        :return: A Deferred that produces an identifier for the stored
+            result.
+        """
+        def to_str(result):
+            return str(result.inserted_id)
+
+        # Store the timestamp field as a special hidden datetime field
+        # for sorting.
+        result['sort$timestamp'] = timestamp_parser.parse(result['timestamp'])
+        id = self.collection.insert_one(result)
+        id.addCallback(to_str)
+        return id
+
+    def retrieve(self, id):
+        """
+        Retrive a result by the given identifier.
+        """
+        try:
+            object_id = ObjectId(id)
+        except InvalidId:
+            raise ResultNotFound()
+
+        def post_process(result):
+            if result is None:
+                raise ResultNotFound()
+            return result
+
+        # Do not include the '_id' and 'sort$timestamp' fields in the
+        # results as these are not part of the original document.
+        # If we later choose to include '_id', its type is 'ObjectId'
+        # which can not be serialized to JSON. Either a custom
+        # JSONEncoder or bson.json_util.dumps would be needed.
+        d = self.collection.find_one(
+            filter={'_id': object_id},
+            fields={'_id': False, 'sort$timestamp': False}
+        )
+        d.addCallback(post_process)
+        return d
+
+    def query(self, filter, limit=None):
+        """
+        Return matching results.
+        """
+        if limit == 0:
+            return succeed([])
+
+        # The txmongo API differs from pymongo with regard to sorting.
+        # To sort results when making a query using txmongo, a query
+        # filter needs to be created and passed to collection.find().
+        sort_filter = orderby(DESCENDING('sort$timestamp'))
+
+        # Do not include the '_id' and 'sort$timestamp' fields in the
+        # results as these are not part of the original document.
+        find_args = dict(
+            filter=sort_filter,
+            fields={'_id': False, 'sort$timestamp': False}
+        )
+        if limit:
+            find_args['limit'] = limit
+
+        return self.collection.find(filter, **find_args)
+
+    def delete(self, id):
+        """
+        Delete a result by the given identifier.
+        """
+        try:
+            object_id = ObjectId(id)
+        except InvalidId:
+            raise ResultNotFound()
```

Comment:
Similar comments to above about InvalidId and a more informative ResultNotFound.